src/testdrive/src/action/sql.rs

```diff
@@ -312,11 +313,27 @@ impl SqlAction {
 }
 
 pub struct FailSqlAction {
-    cmd: FailSqlCommand,
+    query: String,
+    expected_error: ErrorMatcher,
     stmt: Option<Statement<Raw>>,
     context: Context,
 }
 
+enum ErrorMatcher {
+    Contains(String),
+    Exact(String),
+    Regex(Regex),
+}
+
+impl ErrorMatcher {
+    fn as_str(&self) -> &str {
+        match self {
+            ErrorMatcher::Contains(s) | ErrorMatcher::Exact(s) => s.as_str(),
+            ErrorMatcher::Regex(r) => r.as_str(),
+        }
+    }
+}
```

Comment:
Unfortunately variable substitution happens before the parts get passed in here, not at parse time, so that wouldn't work correctly.

Reply:
Hm, good point. It's probably not entirely sensible to be substituting vars into the regex period since they won't get their metachars escaped, but we can address in a followup.